OPTIONS response should include Allow header

kellymclaughlin opened the pull request:
When an `OPTIONS` request is made the response should include an `Allow` header that specifies all the allowed HTTP methods. Also given that there's no content in the response, a `204 No Content` is probably a more appropriate status code.

kellymclaughlin committed:
Return Allow header in response to OPTIONS request

kellymclaughlin commented:
Here's some sample output of modifying the root resource in the airship example to accept `OPTIONS` requests:

```
17:10:00:~ $ curl -v -X OPTIONS http://localhost:3000
* Rebuilt URL to: http://localhost:3000/
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3000 (#0)
> OPTIONS / HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Transfer-Encoding: chunked
< Date: Thu, 30 Jul 2015 23:10:23 GMT
< Server: Warp/3.0.13.1
< Allow: GET,HEAD,OPTIONS
< Airship-Trace: b13,b12,b11,b10,b09,b08,b07,b06,b05,b04,b03
< Airship-Quip: WARNING: ulimit -n is 1024
< 
* Connection #0 to host localhost left intact
```

reiddraper commented on src/Airship/Internal/Decision.hs:
```diff
@@ -242,8 +242,11 @@ b04 r@Resource{..} = do
-        then lift $ halt HTTP.status200
+        then do
+            lift $ addResponseHeader ("Allow",  intercalate "," allowed)
+            lift $ halt HTTP.status200
```
should we use HTTP.status204 here?

dkubb replied:
I almost always return a 204 when there's no response body.

kellymclaughlin replied:
I don't really know tbh. The spec is pretty vague. All it says is:

```
A 200 response SHOULD include any header fields that indicate
   optional features implemented by the server and applicable to that
   resource (e.g., Allow)...
```

I found all of two sites in the wild from a quick and unscientific test that accept OPTIONS requests, but they both return 200. That of course doesn't mean it's correct and given the language in the spec it sounds like we have some leeway so I'm open to 200 or 204:

```
 curl -v -X OPTIONS http://www.reddit.com
* Rebuilt URL to: http://www.reddit.com/
*   Trying 198.41.209.139...
* Connected to www.reddit.com (198.41.209.139) port 80 (#0)
> OPTIONS / HTTP/1.1
> Host: www.reddit.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Fri, 31 Jul 2015 16:47:45 GMT
< Content-Type: text/html; charset=UTF-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< Set-Cookie: __cfduid=d67c1feb7091b0f82a1b087f05630ee5e1438361265; expires=Sat, 30-Jul-16 16:47:45 GMT; path=/; domain=.reddit.com; HttpOnly
< x-ua-compatible: IE=edge
< x-frame-options: SAMEORIGIN
< x-content-type-options: nosniff
< x-xss-protection: 1; mode=block
< cache-control: max-age=0, must-revalidate
< X-Moose: majestic
< Server: cloudflare-nginx
< CF-RAY: 20eac9731a84202a-DFW
< 
* Connection #0 to host www.reddit.com left intact
```

```
 curl -v -X OPTIONS http://www.cnn.com
* Rebuilt URL to: http://www.cnn.com/
*   Trying 23.235.44.73...
* Connected to www.cnn.com (23.235.44.73) port 80 (#0)
> OPTIONS / HTTP/1.1
> Host: www.cnn.com
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< x-servedByHost: prd-10-60-170-45.nodes.56m.dmtio.net
< Cache-Control: max-age=60
< Allow: GET,HEAD
< Content-Type: text/html; charset=utf-8
< Accept-Ranges: bytes
< Via: 1.1 varnish
< Content-Length: 8
< Accept-Ranges: bytes
< Date: Fri, 31 Jul 2015 16:50:49 GMT
< Via: 1.1 varnish
< Connection: keep-alive
< X-Served-By: cache-iad2151-IAD, cache-dfw1827-DFW
< X-Cache: MISS, MISS
< X-Cache-Hits: 0, 0
< X-Timer: S1438361449.638794,VS0,VE65
< Vary: Accept-Encoding
< 
* Connection #0 to host www.cnn.com left intact
GET,HEAD
```

kellymclaughlin replied:
Maybe 204 is best unless we also want to change OPTIONS requests to return a body as the cnn.com request does.

dkubb replied:
> Maybe 204 is best unless we also want to change OPTIONS requests to return a body as the cnn.com request does.

I agree. I use 204 to signal my intention that the client doesn't even have to test the body because it's going to be empty.

There are lots of "RESTful apps" (note the air quotes) that return 200 for _every_ successful response, rather than a more precise status like 201 or 204. I wouldn't necessarily give much weight to in-the-wild behaviour compared to the intended semantics specified in the RFCs.

kellymclaughlin replied:
:+1: Makes sense to me. I'll push another commit to make that update.

kellymclaughlin committed:
Return 204 on OPTIONS requests instead of 200

kellymclaughlin committed:
Address redundant import warning in Decision

Address redundant import warning in the Decision module due to the
ByteString constructor from Data.ByteString being imported twice.
